<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>metagene2: a package to produce metagene plots • metagene2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="metagene2: a package to produce metagene plots">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">metagene2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/metagene2.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>metagene2: a package to produce metagene plots</h1>
            
      
      
      <div class="hidden name"><code>metagene2.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><p><strong>Package</strong>: <em><a href="https://bioconductor.org/packages/3.10/metagene2">metagene2</a></em><br><strong>Modified</strong>: April 2nd, 2019<br><strong>Compiled</strong>: Thu Oct 03 09:45:28 2019<br><strong>License</strong>: Artistic-2.0<br></p>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This package produces metagene plots, and is the successor to the <code>metagene</code> package. Users of <code>metagene</code> can find a list of differences between <code>metagene2</code> and <code>metagene</code> in the <a href="#diff_v1">Differences with metagene</a> section of this vignette.</p>
<p>Metagene plots aggregate coverages from multiple sources (bam files) over multiple regions (genes, cofactor binding sites, etc.) to provide profiles of average coverage. They are useful for many different purposes, such as comparing the binding profiles of DNA-interacting proteins at selected groups of features. In a typical analysis, these features will be the transcription start sites (TSS) of genes, transcription factor binding sites, or enhancer regions. Multiple combinations of groups of features and/or groups of bam files can be compared in a single analysis. The metagene2 package uses bootstrap analysis to provide an estimation of the mean enrichment and a confidence interval for each group of samples.</p>
<p>This vignette will introduce the main features of the metagene2 package. You can load the metagene2 package by calling <code><a href="https://rdrr.io/r/base/library.html">library(metagene2)</a></code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(metagene2)</code></pre></div>
<pre><code>## Warning: package 'GenomicRanges' was built under R version 3.6.1</code></pre>
<pre><code>## Warning: package 'S4Vectors' was built under R version 3.6.1</code></pre>
<pre><code>## Warning: package 'IRanges' was built under R version 3.6.1</code></pre>
<pre><code>## Warning: package 'BiocParallel' was built under R version 3.6.1</code></pre>
</div>
<div id="creating-a-metagene-object" class="section level1">
<h1 class="hasAnchor">
<a href="#creating-a-metagene-object" class="anchor"></a>Creating a metagene object</h1>
<p><code>metagene2</code> objects are used to perform all of the analysis steps necessary to produce metagene plots. Calling <code>metagene2$new</code> creates a <code>metagene2</code> object and requires only two mandatory parameters: <code>bam_files</code>, which is the list of bam files from which coverages should be extracted, and <code>regions</code>, which is the list of regions over which said coverages are computed. We also recommend using the optional <code>assay</code> parameter, which can be one of <code>'chipseq'</code> or <code>'rnaseq'</code>, and will automatically set other optional parameters to convenient defaults. We discuss each of these arguments below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># metagene objects are created by calling metagene2$new and providing</span>
<span class="co"># regions and bam files:</span>
mg &lt;-<span class="st"> </span>metagene2$<span class="kw">new</span>(<span class="dt">regions =</span> <span class="kw"><a href="../reference/get_demo_regions.html">get_demo_regions</a></span>(), 
                   <span class="dt">bam_files =</span> <span class="kw"><a href="../reference/get_demo_bam_files.html">get_demo_bam_files</a></span>(), 
                   <span class="dt">assay=</span><span class="st">'chipseq'</span>)

<span class="co"># We can then plot coverage over those regions across all bam files.</span>
mg$<span class="kw">produce_metagene</span>(<span class="dt">title =</span> <span class="st">"Demo metagene plot"</span>)</code></pre></div>
<p><img src="metagene2_files/figure-html/minimalAnalysis-1.png" width="700"></p>
<div id="specifying-alignment-files-bam-files" class="section level2">
<h2 class="hasAnchor">
<a href="#specifying-alignment-files-bam-files" class="anchor"></a>Specifying alignment files (BAM files)</h2>
<p>There is no hard limit on the number of BAM files that can be included in an analysis. However, loading a large number of bam files might also require large amounts of memory. The provided bam files must be indexed: a file named <code>file.bam</code>, must have an accompanying <code>file.bam.bai</code> or <code>file.bai</code> in its directory.</p>
<p>The paths (relative or absolute) to the BAM files must be provided in a vector. If the vector is named, then those names will be used to refer to the bam files in subsequent steps. Otherwise, <code>metagene2</code> will attempt to generate appropriate names.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># We create a vector with paths to the bam files of interest.</span>
bam_files &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_demo_bam_files.html">get_demo_bam_files</a></span>()
<span class="kw"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(bam_files)</code></pre></div>
<pre><code>## [1] "align1_rep1.bam" "align1_rep2.bam" "align2_rep1.bam" "align2_rep2.bam"
## [5] "ctrl.bam"</code></pre>
<p>Each bam file must have a corresponding index file:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># List .bai matching our specified bam files.</span>
<span class="kw"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Sys.glob.html">Sys.glob</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">".bam"</span>, <span class="st">".ba*"</span>, bam_files)))</code></pre></div>
<pre><code>##  [1] "align1_rep1.bam"     "align1_rep1.bam.bai" "align1_rep2.bam"    
##  [4] "align1_rep2.bam.bai" "align2_rep1.bam"     "align2_rep1.bam.bai"
##  [7] "align2_rep2.bam"     "align2_rep2.bam.bai" "ctrl.bam"           
## [10] "ctrl.bam.bai"</code></pre>
<p>If no names were provided for the bam files, metagene automatically generates some:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mg &lt;-<span class="st"> </span>metagene2$<span class="kw">new</span>(<span class="dt">regions =</span> <span class="kw"><a href="../reference/get_demo_regions.html">get_demo_regions</a></span>(), <span class="dt">bam_files =</span> bam_files)
<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(mg$<span class="kw">get_params</span>()[[<span class="st">"bam_files"</span>]])</code></pre></div>
<pre><code>## [1] "align1_rep1" "align1_rep2" "align2_rep1" "align2_rep2" "ctrl"</code></pre>
<p>We also could have explicitly named our bam files.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(bam_files) =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"a1_1"</span>, <span class="st">"a1_2"</span>, <span class="st">"a2_1"</span>, <span class="st">"a2_2"</span>, <span class="st">"ctrl"</span>)
mg &lt;-<span class="st"> </span>metagene2$<span class="kw">new</span>(<span class="dt">regions =</span> <span class="kw"><a href="../reference/get_demo_regions.html">get_demo_regions</a></span>(), <span class="dt">bam_files =</span> bam_files)
<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(mg$<span class="kw">get_params</span>()[[<span class="st">"bam_files"</span>]])</code></pre></div>
<pre><code>## [1] "a1_1" "a1_2" "a2_1" "a2_2" "ctrl"</code></pre>
</div>
<div id="specifying-genomic-regions" class="section level2">
<h2 class="hasAnchor">
<a href="#specifying-genomic-regions" class="anchor"></a>Specifying genomic regions</h2>
<p>The regions for the metagene analysis can be provided in one of three different formats:</p>
<ul>
<li>A <code>character</code> vector, containing the paths to bed, narrowPeak, broadPeak or gtf files describing the regions to be used.</li>
<li>A <code>GRanges</code> or <code>GRangesList</code> object defining a set of contiguous regions.</li>
<li>A <code>GRangesList</code> where each element defines a set of regions to be stitched together to be considered as a single logical region.</li>
</ul>
<div id="defining-regions-using-bed-narrowpeak-broadpeak-and-gtf-files" class="section level3">
<h3 class="hasAnchor">
<a href="#defining-regions-using-bed-narrowpeak-broadpeak-and-gtf-files" class="anchor"></a>Defining regions using BED, narrowPeak, broadPeak and GTF files</h3>
<p><code>metagene2</code> can automatically import your regions of interest if they are already defined in a file with one of the following formats:</p>
<ul>
<li><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">bed</a></li>
<li><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format12">narrowPeak</a></li>
<li><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format13">broadPeak</a></li>
<li><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format4">gtf</a></li>
</ul>
<p>A file’s extension will usually reflect the format it is stored in.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">regions &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_demo_region_filenames.html">get_demo_region_filenames</a></span>()
regions</code></pre></div>
<pre><code>## [1] "C:/Program Files/R/R-3.6.0beta/library/metagene2/extdata/list1.bed"
## [2] "C:/Program Files/R/R-3.6.0beta/library/metagene2/extdata/list2.bed"</code></pre>
<p>By providing those two file names to <code>metagene2$new</code>, they will be loaded and converted into appropriate objects:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mg &lt;-<span class="st"> </span>metagene2$<span class="kw">new</span>(<span class="dt">regions =</span> <span class="kw"><a href="../reference/get_demo_region_filenames.html">get_demo_region_filenames</a></span>(),
                   <span class="dt">bam_files =</span> <span class="kw"><a href="../reference/get_demo_bam_files.html">get_demo_bam_files</a></span>())
mg$<span class="kw">get_regions</span>()</code></pre></div>
<pre><code>## GRanges object with 100 ranges and 3 metadata columns:
##         seqnames            ranges strand |        name     score
##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;
##     [1]     chr1 16103663-16105662      + |     list1_1         0
##     [2]     chr1 23921318-23923317      + |     list1_2         0
##     [3]     chr1 34848977-34850976      + |     list1_3         0
##     [4]     chr1 36368182-36370181      + |     list1_4         0
##     [5]     chr1 36690488-36692487      + |     list1_5         0
##     ...      ...               ...    ... .         ...       ...
##    [96]     chr1 81075951-81077950      - |    list2_46         0
##    [97]     chr1 85108854-85110853      - |    list2_47         0
##    [98]     chr1 85960056-85962055      - |    list2_48         0
##    [99]     chr1 86110971-86112970      - |    list2_49         0
##   [100]     chr1 87155522-87157521      - |    list2_50         0
##         region_name
##         &lt;character&gt;
##     [1]       list1
##     [2]       list1
##     [3]       list1
##     [4]       list1
##     [5]       list1
##     ...         ...
##    [96]       list2
##    [97]       list2
##    [98]       list2
##    [99]       list2
##   [100]       list2
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="defining-contiguous-regions-using-granges-or-grangeslist-objects" class="section level3">
<h3 class="hasAnchor">
<a href="#defining-contiguous-regions-using-granges-or-grangeslist-objects" class="anchor"></a>Defining contiguous regions using GRanges or GRangesList objects</h3>
<p>As an alternative to a list of BED files, <code>GRanges</code> objects can be used to define contiguous regions of interest. Each range defined within the <code>GRanges</code> object is treated separately from the others. <code>GRangesList</code> objects are also accepted, but they are automatically coerced into <code>GRanges</code> objects, and a column named <code>region_name</code> bearing the name of the list elements is added to the coerced <code>GRanges</code>. Here is an example of valid regions provided as a <code>GRangesList</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">regions &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_demo_regions.html">get_demo_regions</a></span>()
regions</code></pre></div>
<pre><code>## GRangesList object of length 2:
## $list1
## GRanges object with 50 ranges and 2 metadata columns:
##        seqnames              ranges strand |        name     score
##           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;
##    [1]     chr1   16103663-16105662      + |     list1_1         0
##    [2]     chr1   23921318-23923317      + |     list1_2         0
##    [3]     chr1   34848977-34850976      + |     list1_3         0
##    [4]     chr1   36368182-36370181      + |     list1_4         0
##    [5]     chr1   36690488-36692487      + |     list1_5         0
##    ...      ...                 ...    ... .         ...       ...
##   [46]     chr1 172081530-172083529      + |    list1_46         0
##   [47]     chr1 172081796-172083795      + |    list1_47         0
##   [48]     chr1 172147016-172149015      + |    list1_48         0
##   [49]     chr1 172205805-172207804      + |    list1_49         0
##   [50]     chr1 172260642-172262641      + |    list1_50         0
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths
## 
## $list2
## GRanges object with 50 ranges and 2 metadata columns:
##        seqnames            ranges strand |        name     score
##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;
##    [1]     chr1   3670499-3672498      - |     list2_1         0
##    [2]     chr1   5916399-5918398      - |     list2_2         0
##    [3]     chr1   9699210-9701209      - |     list2_3         0
##    [4]     chr1   9907639-9909638      - |     list2_4         0
##    [5]     chr1 10718946-10720945      - |     list2_5         0
##    ...      ...               ...    ... .         ...       ...
##   [46]     chr1 81075951-81077950      - |    list2_46         0
##   [47]     chr1 85108854-85110853      - |    list2_47         0
##   [48]     chr1 85960056-85962055      - |    list2_48         0
##   [49]     chr1 86110971-86112970      - |    list2_49         0
##   [50]     chr1 87155522-87157521      - |    list2_50         0
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>When loaded by <code>metagene2</code>, they are converted to a <code>GRanges</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mg &lt;-<span class="st"> </span>metagene2$<span class="kw">new</span>(<span class="dt">regions =</span> regions,
                   <span class="dt">bam_files =</span> <span class="kw"><a href="../reference/get_demo_bam_files.html">get_demo_bam_files</a></span>())
mg$<span class="kw">get_regions</span>()</code></pre></div>
<pre><code>## GRanges object with 100 ranges and 3 metadata columns:
##         seqnames            ranges strand |        name     score
##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;
##     [1]     chr1 16103663-16105662      + |     list1_1         0
##     [2]     chr1 23921318-23923317      + |     list1_2         0
##     [3]     chr1 34848977-34850976      + |     list1_3         0
##     [4]     chr1 36368182-36370181      + |     list1_4         0
##     [5]     chr1 36690488-36692487      + |     list1_5         0
##     ...      ...               ...    ... .         ...       ...
##    [96]     chr1 81075951-81077950      - |    list2_46         0
##    [97]     chr1 85108854-85110853      - |    list2_47         0
##    [98]     chr1 85960056-85962055      - |    list2_48         0
##    [99]     chr1 86110971-86112970      - |    list2_49         0
##   [100]     chr1 87155522-87157521      - |    list2_50         0
##         region_name
##         &lt;character&gt;
##     [1]       list1
##     [2]       list1
##     [3]       list1
##     [4]       list1
##     [5]       list1
##     ...         ...
##    [96]       list2
##    [97]       list2
##    [98]       list2
##    [99]       list2
##   [100]       list2
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>For more details about each datasets, please refer to their documentation (i.e.:<code>?promoters_hg19</code>).</p>
</div>
<div id="grangeslist-objects-for-stitching-ranges-together" class="section level3">
<h3 class="hasAnchor">
<a href="#grangeslist-objects-for-stitching-ranges-together" class="anchor"></a>GRangesList objects for stitching ranges together</h3>
<p>For certain types of analyses, it is useful to stitch together several regions into one logical unit. This is the case in RNA-seq data, where exons are individual regions which make more sense when grouped together into a single transcript.</p>
<p>For these cases, <code>regions</code> can be a <code>GRangesList</code> object where each element is one such logical region. One must also specify the <code>region_mode="stitch"</code> parameter when creating the new metagene object. When <code>assay='rnaseq'</code>, <code>region_mode</code> is automatically set to <code>"stitch"</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">regions &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_demo_rna_regions.html">get_demo_rna_regions</a></span>()
regions</code></pre></div>
<pre><code>## GRangesList object of length 2:
## $DPM1
## GRanges object with 10 ranges and 2 metadata columns:
##        seqnames            ranges strand |        name     score
##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;
##    [1]    chr20 50934868-50935236      - |        &lt;NA&gt;      &lt;NA&gt;
##    [2]    chr20 50936149-50936262      - |        &lt;NA&gt;      &lt;NA&gt;
##    [3]    chr20 50940866-50940955      - |        &lt;NA&gt;      &lt;NA&gt;
##    [4]    chr20 50941106-50941209      - |        &lt;NA&gt;      &lt;NA&gt;
##    [5]    chr20 50942032-50942126      - |        &lt;NA&gt;      &lt;NA&gt;
##    [6]    chr20 50945738-50945762      - |        &lt;NA&gt;      &lt;NA&gt;
##    [7]    chr20 50945848-50945923      - |        &lt;NA&gt;      &lt;NA&gt;
##    [8]    chr20 50948630-50948662      - |        &lt;NA&gt;      &lt;NA&gt;
##    [9]    chr20 50955187-50955285      - |        &lt;NA&gt;      &lt;NA&gt;
##   [10]    chr20 50958364-50958555      - |        &lt;NA&gt;      &lt;NA&gt;
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths
## 
## $NDUFAB1
## GRanges object with 7 ranges and 2 metadata columns:
##       seqnames            ranges strand |        name     score
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;
##   [1]    chr16 23581003-23581173      - |        &lt;NA&gt;      &lt;NA&gt;
##   [2]    chr16 23581872-23582375      - |        &lt;NA&gt;      &lt;NA&gt;
##   [3]    chr16 23585337-23585423      - |        &lt;NA&gt;      &lt;NA&gt;
##   [4]    chr16 23587198-23587319      - |        &lt;NA&gt;      &lt;NA&gt;
##   [5]    chr16 23590887-23591153      - |        &lt;NA&gt;      &lt;NA&gt;
##   [6]    chr16 23595430-23595638      - |        &lt;NA&gt;      &lt;NA&gt;
##   [7]    chr16 23596124-23596356      - |        &lt;NA&gt;      &lt;NA&gt;
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>In stitch mode, the loaded regions remain in a <code>GRangesList</code>, rather than being coerced into a <code>GRanges</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mg &lt;-<span class="st"> </span>metagene2$<span class="kw">new</span>(<span class="dt">regions =</span> regions,
                   <span class="dt">bam_files =</span> <span class="kw"><a href="../reference/get_demo_rna_bam_files.html">get_demo_rna_bam_files</a></span>(),
                   <span class="dt">region_mode=</span><span class="st">"stitch"</span>)
mg$<span class="kw">get_regions</span>()</code></pre></div>
<pre><code>## GRangesList object of length 2:
## $DPM1
## GRanges object with 10 ranges and 3 metadata columns:
##        seqnames            ranges strand |        name     score
##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;
##    [1]    chr20 50934868-50935236      - |        &lt;NA&gt;      &lt;NA&gt;
##    [2]    chr20 50936149-50936262      - |        &lt;NA&gt;      &lt;NA&gt;
##    [3]    chr20 50940866-50940955      - |        &lt;NA&gt;      &lt;NA&gt;
##    [4]    chr20 50941106-50941209      - |        &lt;NA&gt;      &lt;NA&gt;
##    [5]    chr20 50942032-50942126      - |        &lt;NA&gt;      &lt;NA&gt;
##    [6]    chr20 50945738-50945762      - |        &lt;NA&gt;      &lt;NA&gt;
##    [7]    chr20 50945848-50945923      - |        &lt;NA&gt;      &lt;NA&gt;
##    [8]    chr20 50948630-50948662      - |        &lt;NA&gt;      &lt;NA&gt;
##    [9]    chr20 50955187-50955285      - |        &lt;NA&gt;      &lt;NA&gt;
##   [10]    chr20 50958364-50958555      - |        &lt;NA&gt;      &lt;NA&gt;
##        region_name
##        &lt;character&gt;
##    [1]        DPM1
##    [2]        DPM1
##    [3]        DPM1
##    [4]        DPM1
##    [5]        DPM1
##    [6]        DPM1
##    [7]        DPM1
##    [8]        DPM1
##    [9]        DPM1
##   [10]        DPM1
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths
## 
## $NDUFAB1
## GRanges object with 7 ranges and 3 metadata columns:
##       seqnames            ranges strand |        name     score
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;
##   [1]    chr16 23581003-23581173      - |        &lt;NA&gt;      &lt;NA&gt;
##   [2]    chr16 23581872-23582375      - |        &lt;NA&gt;      &lt;NA&gt;
##   [3]    chr16 23585337-23585423      - |        &lt;NA&gt;      &lt;NA&gt;
##   [4]    chr16 23587198-23587319      - |        &lt;NA&gt;      &lt;NA&gt;
##   [5]    chr16 23590887-23591153      - |        &lt;NA&gt;      &lt;NA&gt;
##   [6]    chr16 23595430-23595638      - |        &lt;NA&gt;      &lt;NA&gt;
##   [7]    chr16 23596124-23596356      - |        &lt;NA&gt;      &lt;NA&gt;
##       region_name
##       &lt;character&gt;
##   [1]     NDUFAB1
##   [2]     NDUFAB1
##   [3]     NDUFAB1
##   [4]     NDUFAB1
##   [5]     NDUFAB1
##   [6]     NDUFAB1
##   [7]     NDUFAB1
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="common_ranges" class="section level3">
<h3 class="hasAnchor">
<a href="#common_ranges" class="anchor"></a>Generating common ranges (Promoters, gene bodies)</h3>
<p>Some common ranges that can be useful for plotting include the set of all TSSes or gene bodies. While metagene2 does not provide those, they can easily be generated using packages from BioConductor:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># First locate the TxDb package containing the geneset of interest.</span>
    <span class="co"># Some of the most common TxDb packages include:</span>
    <span class="co">#  - TxDb.Hsapiens.UCSC.hg38.knownGene</span>
    <span class="co">#  - TxDb.Hsapiens.UCSC.hg19.knownGene</span>
    <span class="co">#  - TxDb.Mmusculus.UCSC.mm10.knownGene</span>
    <span class="co">#  - TxDb.Mmusculus.UCSC.mm10.ensGene</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(TxDb.Hsapiens.UCSC.hg38.knownGene)
    
    <span class="co"># We'll use the GenomicFeatures package to obtain gene/TSS coordinates</span>
    <span class="co"># from the TxDb package.</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(GenomicFeatures)
    
    <span class="co"># The GenomicFeatures::genes function provides us with gene bodies.</span>
    all_gene_bodies =<span class="st"> </span>GenomicFeatures::<span class="kw">genes</span>(TxDb.Hsapiens.UCSC.hg38.knownGene)
    
    <span class="co"># The GenomicFeatures::promoters function gets a region flanking the TSS.</span>
    <span class="co"># By using it directly on TxDb.Hsapiens.UCSC.hg38.knownGene, we would get</span>
    <span class="co"># the TSSes of all transcripts. Here, we use it on the gene_bodies GRanges</span>
    <span class="co"># we've just created, and limit ourselves to one TSS per gene.</span>
    all_TSS =<span class="st"> </span>GenomicFeatures::<span class="kw">promoters</span>(all_gene_bodies,
                                         <span class="dt">upstream=</span><span class="dv">2000</span>, <span class="dt">downstream=</span><span class="dv">2000</span>)</code></pre></div>
</div>
</div>
</div>
<div id="grouping-regions-and-bam-files" class="section level1">
<h1 class="hasAnchor">
<a href="#grouping-regions-and-bam-files" class="anchor"></a>Grouping regions and bam files</h1>
<p>By default, <code>metagene2</code> aggregates all passed-in regions together, and treats all bam files separately. However, most non-trivial analyses will benefit from more granularity. Bam files can be split among different ChIP-seq experiments and/or multiple replicates. Regions can likewise be split according to multiple criteria: is the underlying gene up- or down-regulated? Is the enhancer bound by a cofactor of interest? Below, we discuss how <code>metagene2</code> allows the user to specify those groupings to produce relevant analyses.</p>
<div id="grouping-bam-files" class="section level2">
<h2 class="hasAnchor">
<a href="#grouping-bam-files" class="anchor"></a>Grouping bam files</h2>
<div id="using-an-experimental-design" class="section level3">
<h3 class="hasAnchor">
<a href="#using-an-experimental-design" class="anchor"></a>Using an experimental design</h3>
<p>In <code>metagene2</code>, an experimental design is a set of design groups, each of which is defined as a set of “input” bam files and a set of “control” bam files. There is no limit to the number of design groups, though a large number of design groups will require a proportionately large amount of memory. A BAM file can be assigned to more than one design group.</p>
<p>The experimental design is expressed using a data-frame, where each row represents a bam file. The very first column of the data-frame must identify the bam files, using either their paths or their names as specified in the <code>bam_files</code> argument. Each subsequent column then represents an individual design group. The column name defines the design group’s name, and the column values determine how each bam file relates to the design group:</p>
<pre><code>* 0: ignore file
* 1: input
* 2: control</code></pre>
<p>A design group does not need to have a control, but it must have at least one input. Control samples are ignored when no normalization or “RPM” normalization is chosen. However, they are used to remove background noise using “NCIS” normalization is selected, or to compute coverage ratios with a control sample when “log2_ratio” normalization is applied.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">example_design &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">Samples =</span> bam_files,
                             <span class="dt">align1 =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">2</span>),
                             <span class="dt">align2 =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">kable</span>(example_design)</code></pre></div>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="left">Samples</th>
<th align="right">align1</th>
<th align="right">align2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>a1_1</td>
<td align="left">C:/Program Files/R/R-3.6.0beta/library/metagene2/extdata/align1_rep1.bam</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td>a1_2</td>
<td align="left">C:/Program Files/R/R-3.6.0beta/library/metagene2/extdata/align1_rep2.bam</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td>a2_1</td>
<td align="left">C:/Program Files/R/R-3.6.0beta/library/metagene2/extdata/align2_rep1.bam</td>
<td align="right">0</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td>a2_2</td>
<td align="left">C:/Program Files/R/R-3.6.0beta/library/metagene2/extdata/align2_rep2.bam</td>
<td align="right">0</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td>ctrl</td>
<td align="left">C:/Program Files/R/R-3.6.0beta/library/metagene2/extdata/ctrl.bam</td>
<td align="right">2</td>
<td align="right">2</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Initializing the metagene object.</span>
mg &lt;-<span class="st"> </span>metagene2$<span class="kw">new</span>(<span class="dt">regions =</span> <span class="kw"><a href="../reference/get_demo_regions.html">get_demo_regions</a></span>(),
                   <span class="dt">bam_files =</span> <span class="kw"><a href="../reference/get_demo_bam_files.html">get_demo_bam_files</a></span>(),
                   <span class="dt">assay=</span><span class="st">'chipseq'</span>)

<span class="co"># Plotting while grouping the bam files by design group</span>
mg$<span class="kw">produce_metagene</span>(<span class="dt">design=</span>example_design)</code></pre></div>
<p><img src="metagene2_files/figure-html/design_plot-1.png" width="700"></p>
</div>
<div id="using-design-metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#using-design-metadata" class="anchor"></a>Using design metadata</h3>
<p>Grouping bam files using an experimental design aggregates all of their coverages together, flattening them into a single mean value and its accompanying confidence interval. In some cases, it might be preferable to keep all experimental replicates separate, and plot them next to each other to assess experimental reproducibility.</p>
<p><code>metagene2</code> allows you to specify metadata to accompany your experimental design, and then allows you to plot your data accordingly:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Initializing the metagene object.</span>
mg &lt;-<span class="st"> </span>metagene2$<span class="kw">new</span>(<span class="dt">regions =</span> <span class="kw"><a href="../reference/get_demo_regions.html">get_demo_regions</a></span>(),
                   <span class="dt">bam_files =</span> <span class="kw"><a href="../reference/get_demo_bam_files.html">get_demo_bam_files</a></span>()[<span class="dv">1</span>:<span class="dv">4</span>],
                   <span class="dt">assay=</span><span class="st">'chipseq'</span>)

design_meta =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">design=</span>mg$<span class="kw">get_design_group_names</span>(),
                         <span class="dt">Align=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Align1"</span>, <span class="st">"Align1"</span>, <span class="st">"Align2"</span>, <span class="st">"Align2"</span>),
                         <span class="dt">Rep=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">2</span>))
                             
mg$<span class="kw">produce_metagene</span>(<span class="dt">design_metadata=</span>design_meta, <span class="dt">facet_by=</span>Align~Rep, <span class="dt">group_by=</span><span class="st">"region"</span>)</code></pre></div>
<pre><code>## Warning: Column `design` joining character vector and factor, coercing into
## character vector</code></pre>
<p><img src="metagene2_files/figure-html/design_metadata-1.png" width="700"></p>
</div>
</div>
<div id="grouping-regions" class="section level2">
<h2 class="hasAnchor">
<a href="#grouping-regions" class="anchor"></a>Grouping regions</h2>
<p>The descriptive power of metagenes stem from their ability to succintly summarize coverage over groups of regions with shared characteristics. The <code>metagene2</code> package provides two options for grouping regions together: explicit grouping using a GRangesList object, or grouping using metadata.</p>
<div id="grouping-regions-using-a-grangeslist" class="section level3">
<h3 class="hasAnchor">
<a href="#grouping-regions-using-a-grangeslist" class="anchor"></a>Grouping regions using a GRangesList</h3>
<p>When working with separate, contiguous regions, the most straightforward way of grouping regions together is to pass a GRangesList instead of a GRanges to metagene2$new. Each element of the list then becomes a group of region that metagene aggregates together. The same behaviour is obtained if regions are specified through file names rather than GRanges objects.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Create a GRangesList of regions to be grouped together.</span>
regions_grl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_demo_regions.html">get_demo_regions</a></span>()

<span class="co"># We now have a named GRangesList with two set of 50 regions.</span>
regions_grl</code></pre></div>
<pre><code>## GRangesList object of length 2:
## $list1
## GRanges object with 50 ranges and 2 metadata columns:
##        seqnames              ranges strand |        name     score
##           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;
##    [1]     chr1   16103663-16105662      + |     list1_1         0
##    [2]     chr1   23921318-23923317      + |     list1_2         0
##    [3]     chr1   34848977-34850976      + |     list1_3         0
##    [4]     chr1   36368182-36370181      + |     list1_4         0
##    [5]     chr1   36690488-36692487      + |     list1_5         0
##    ...      ...                 ...    ... .         ...       ...
##   [46]     chr1 172081530-172083529      + |    list1_46         0
##   [47]     chr1 172081796-172083795      + |    list1_47         0
##   [48]     chr1 172147016-172149015      + |    list1_48         0
##   [49]     chr1 172205805-172207804      + |    list1_49         0
##   [50]     chr1 172260642-172262641      + |    list1_50         0
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths
## 
## $list2
## GRanges object with 50 ranges and 2 metadata columns:
##        seqnames            ranges strand |        name     score
##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;
##    [1]     chr1   3670499-3672498      - |     list2_1         0
##    [2]     chr1   5916399-5918398      - |     list2_2         0
##    [3]     chr1   9699210-9701209      - |     list2_3         0
##    [4]     chr1   9907639-9909638      - |     list2_4         0
##    [5]     chr1 10718946-10720945      - |     list2_5         0
##    ...      ...               ...    ... .         ...       ...
##   [46]     chr1 81075951-81077950      - |    list2_46         0
##   [47]     chr1 85108854-85110853      - |    list2_47         0
##   [48]     chr1 85960056-85962055      - |    list2_48         0
##   [49]     chr1 86110971-86112970      - |    list2_49         0
##   [50]     chr1 87155522-87157521      - |    list2_50         0
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(regions_grl, length)</code></pre></div>
<pre><code>## $list1
## [1] 50
## 
## $list2
## [1] 50</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Initializing the metagene object.</span>
mg &lt;-<span class="st"> </span>metagene2$<span class="kw">new</span>(<span class="dt">regions =</span> regions_grl,
                   <span class="dt">bam_files =</span> <span class="kw"><a href="../reference/get_demo_bam_files.html">get_demo_bam_files</a></span>(),
                   <span class="dt">assay=</span><span class="st">'chipseq'</span>)

<span class="co"># When plotting the final metagene, our regions are grouped according to</span>
<span class="co"># their membership in the initial GRangesList object.</span>
mg$<span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="dt">facet_by=</span>~region, <span class="dt">group_by=</span><span class="st">"design"</span>)</code></pre></div>
<p><img src="metagene2_files/figure-html/group_region_grangeslist-1.png" width="700"></p>
</div>
<div id="grouping-regions-using-metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#grouping-regions-using-metadata" class="anchor"></a>Grouping regions using metadata</h3>
<p>A more powerful and flexible way of grouping regions is providing region metadata. This is done by using the region_metadata parameter, which must be a data-frame with as many rows as there are regions and columns that can be converted to factors. Once metadata has been attached to regions, they can be grouped using the split_by parameter, which must be a vector of column names from the region_metadata parameter.</p>
<p>By default, if no region_metadata is specified, <code>metagene2</code> looks for the mcols attribute of the <code>regions</code> parameter if it is a <code>GRanges</code> object, or the first line of each individual mcols if it is a <code>GRangesList</code> object.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># First, we load the regions.</span>
regions_gr &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="kw"><a href="../reference/get_demo_regions.html">get_demo_regions</a></span>())

<span class="co"># We then define some metadata.</span>
<span class="co"># The examples here are nonsensical. Real metadata could include factor</span>
<span class="co"># binding status, differential expression, etc.</span>
demo_metadata =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">BedName=</span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(regions_gr),
                           <span class="dt">EvenStart=</span><span class="kw"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>((<span class="kw"><a href="https://rdrr.io/r/stats/start.html">start</a></span>(regions_gr) %%<span class="st"> </span><span class="dv">2</span>) ==<span class="st"> </span><span class="dv">0</span>, <span class="st">"Even"</span>, <span class="st">"Odd"</span>),
                           <span class="dt">Strand=</span><span class="kw">strand</span>(regions_gr))
<span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(demo_metadata)</code></pre></div>
<pre><code>##   BedName EvenStart Strand
## 1   list1       Odd      +
## 2   list1      Even      +
## 3   list1       Odd      +
## 4   list1      Even      +
## 5   list1      Even      +
## 6   list1       Odd      +</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Initializing the metagene object, passing in region metadata.</span>
mg &lt;-<span class="st"> </span>metagene2$<span class="kw">new</span>(<span class="dt">regions =</span> <span class="kw"><a href="../reference/get_demo_regions.html">get_demo_regions</a></span>(),
                   <span class="dt">region_metadata=</span>demo_metadata,
                   <span class="dt">bam_files =</span> <span class="kw"><a href="../reference/get_demo_bam_files.html">get_demo_bam_files</a></span>(),
                   <span class="dt">assay=</span><span class="st">'chipseq'</span>)

<span class="co"># When plotting the metagene, our regions are grouped according to</span>
<span class="co"># the specified metadata columns.</span>
mg$<span class="kw">produce_metagene</span>(<span class="dt">split_by=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"EvenStart"</span>, <span class="st">"Strand"</span>), <span class="dt">facet_by=</span>EvenStart~Strand, <span class="dt">group_by=</span><span class="st">"design"</span>)</code></pre></div>
<p><img src="metagene2_files/figure-html/group_region_metadata-1.png" width="700"></p>
</div>
</div>
</div>
<div id="intermediary_steps" class="section level1">
<h1 class="hasAnchor">
<a href="#intermediary_steps" class="anchor"></a>Intermediary processing steps and further parameters</h1>
<p>A full metagene analysis consists of several steps, each of which produces an intermediary result of interest. Calling the <code>new</code> and <code>produce_metagene</code> methods automatically perform all of those steps sequentially. However, if a full analysis is not required, it is also possible to carry these out one by one. When calling any step of the chain, all previous steps are automatically carried out if they hadn’t previously been. The following schema illustrates those intermediate steps and results: <img src="img/processing_step_schema.png" alt="schematic representation of processing steps"></p>
<div id="arguments-results-caching-and-chaining" class="section level2">
<h2 class="hasAnchor">
<a href="#arguments-results-caching-and-chaining" class="anchor"></a>Arguments, results caching and chaining</h2>
<p>Objects of the <code>metagene2</code> class are “pipeline” objects, whose primary purpose is to chain together the various steps required to produce a metagene plot and manage the parameters required to do so. As such, metagene maintains an internal list of all analytical parameters, which can be obtained by calling <code>get_params()</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mg &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_demo_metagene.html">get_demo_metagene</a></span>()
<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(mg$<span class="kw">get_params</span>())</code></pre></div>
<pre><code>##  [1] "design"                 "design_metadata"       
##  [3] "bam_files"              "padding_size"          
##  [5] "verbose"                "force_seqlevels"       
##  [7] "paired_end"             "assay"                 
##  [9] "strand_specific"        "paired_end_strand_mode"
## [11] "normalization"          "avoid_gaps"            
## [13] "gaps_threshold"         "bin_count"             
## [15] "alpha"                  "sample_count"          
## [17] "region_mode"            "split_by"              
## [19] "region_filter"          "design_filter"         
## [21] "resampling_strategy"    "facet_by"              
## [23] "group_by"               "title"                 
## [25] "x_label"                "extend_reads"          
## [27] "invert_strand"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mg$<span class="kw">get_params</span>()[[<span class="st">"bin_count"</span>]]</code></pre></div>
<pre><code>## [1] 100</code></pre>
<p>Any of these parameters can be set when calling <code>metagene2$new</code> or <code>produce_metagene</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mg &lt;-<span class="st"> </span>metagene2$<span class="kw">new</span>(<span class="dt">regions=</span><span class="kw"><a href="../reference/get_demo_regions.html">get_demo_regions</a></span>(), 
                   <span class="dt">bam_files=</span><span class="kw"><a href="../reference/get_demo_bam_files.html">get_demo_bam_files</a></span>(),
                   <span class="dt">bin_count=</span><span class="dv">50</span>)
mg$<span class="kw">produce_metagene</span>(<span class="dt">alpha=</span><span class="fl">0.01</span>, <span class="dt">title=</span><span class="st">"Set parameters on produce_metagene"</span>)</code></pre></div>
<p><img src="metagene2_files/figure-html/setParamsConstructor-1.png" width="700"></p>
<p>These parameter values can be overwritten in further calls to produce_metagene. All parameters for metagene’s intermediary steps default to NA, which means “keep the previous value for this parameter”. When metagene detects that a parameter has changed, it invalidates only the necessary caches, and updates the parameter value in its internal list.</p>
<p>In the following example, we regenerate the previous metagene object by changing the number of bins. Genome-wide coverages are not computed again, and our previous alpha value and title remain the same.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mg$<span class="kw">produce_metagene</span>(<span class="dt">bin_count=</span><span class="dv">100</span>)</code></pre></div>
<p><img src="metagene2_files/figure-html/changeSingleParamProduceMetagene-1.png" width="700"></p>
<p>Below, we provide a short explanation of all of the above steps, as well as a brief description of their most commonly used parameters. You can refer to each function’s formal documentation in the <code>metagene2</code> manual for more details.</p>
</div>
<div id="metagene2new" class="section level2">
<h2 class="hasAnchor">
<a href="#metagene2new" class="anchor"></a><code>metagene2$new</code>
</h2>
<p>Initializing the metagene object calculates genome wide coverages for all bam files, and performs some preprocessing on the regions of interest. * <strong>regions</strong>: Use this argument to specify which regions the metagene must be computed over. See section 2.2, “Specifying genomic regions”.</p>
<ul>
<li><p><strong>bam_files</strong>: Use this argument to specify which bam_files contain sample information. See section 2.1, “Specifying alignment files”.</p></li>
<li><p><strong>assay</strong>: You can use this as a shorthand for specifying sensible defaults for analysis parameters based on the type of experiment. You can use “chipseq” for a strand agnostic experiment using contiguous regions, or “rnaseq” for a strand-specific experiment where regions are stitched exons.</p></li>
<li><p><strong>region_mode</strong>:Set the way the regions parameter is interpreted. Can be ‘separate’, ‘stitch’ or ‘auto’. See section 2.2, “Specifying regions”.</p></li>
<li><p><strong>padding_size</strong>: The regions defined in <code>regions</code> will be padded by this many nucleotides at both ends.</p></li>
<li><p><strong>cores</strong>: By passing an integer (the number of cores to use) or a <code>BiocParallelParam</code> object, this argument allows metagene to run certain operations in parallel.</p></li>
<li><p><strong>paired_end</strong>: Setting this to true indicates that your bam files containing paired_end data.</p></li>
<li><p><strong>strand_specific</strong>: Set this to TRUE if you want reads on the opposite strand to be discarded when calculating coverages.</p></li>
<li><p><strong>paired_end_strand_mode</strong>: When <code>strand_specific</code> is TRUE, this flag determines how read orientation should be interpreted. See the documentation for <code>GenomicAlignments:::readGAlignmentPairs</code> for the possible values.</p></li>
<li><p><strong>region_metadata</strong>: Use this to specify metadata about your regions. It must have as many rows as you have regions. You can then use the columns therein for the <code>split_by</code> (in <code>mg$split_coverages_by_regions</code>), <code>group_by</code> and <code>facet_by</code> (in <code>mg$plot()</code>) parameters.</p></li>
<li><p><strong>extend_reads</strong>: When calculating coverages, reads will be extended as if they were this long. In single-end chip-seq experiments, the captured fragments are usually longer than their sequenced reads. This option allows for “restoring” those fragments, and provides smoother coverages.</p></li>
<li><p><strong>invert_strand</strong>: Set this to TRUE to invert strands when computing coverages. thisis useful when dealing with single-end stranded RNA-seq, which is often based cDNA rather than mRNAs.</p></li>
</ul>
</div>
<div id="group_coverages" class="section level2">
<h2 class="hasAnchor">
<a href="#group_coverages" class="anchor"></a><code>group_coverages</code>
</h2>
<p>This function groups bam file coverages into design group coverages, and performs normalization and noise removal when requested.</p>
<ul>
<li>
<strong>design</strong>: The design explains how bam files are grouped together into logical design groups. See section 3.1, “Grouping bam files”. By default, each bam file is in its own design group.</li>
<li>
<strong>normalization</strong>: This parameter determines how the coverages are normalized. There are four possible values: “RPM” (Reads per million), “log2_ratio” (log2((input RPM + 1) / (control RPM + 1))), “NCIS” (See Liand and Keles 2012) or NULL (No normalization).</li>
<li>
<strong>design_filter</strong>: You can exclude certain design groups from further processing by filtering them. This is useful if you want to quickly recalculate certain values without processing every single bam file.</li>
</ul>
</div>
<div id="bin_coverages" class="section level2">
<h2 class="hasAnchor">
<a href="#bin_coverages" class="anchor"></a><code>bin_coverages</code>
</h2>
<ul>
<li>
<strong>bin_count</strong>: Determines the number of bins regions will be divided into. <code>bin_count</code> cannot be smaller than the size of your smallest region.</li>
<li>
<strong>region_filter</strong>: You can exclude certain regions from further processing by filtering them. This is useful if you want to quickly recalculate certain values without processing every single region.</li>
</ul>
</div>
<div id="split_coverages_by_regions" class="section level2">
<h2 class="hasAnchor">
<a href="#split_coverages_by_regions" class="anchor"></a><code>split_coverages_by_regions</code>
</h2>
<p>This function splits the coverage matrices into submatrices where all regions have identical metadata.</p>
<ul>
<li>
<strong>split_by</strong>: This parameter is used to group together regions with similar metadata. See section 3.2.2, “Grouping regions using metadata”.</li>
</ul>
</div>
<div id="calculate_ci" class="section level2">
<h2 class="hasAnchor">
<a href="#calculate_ci" class="anchor"></a><code>calculate_ci</code>
</h2>
<p><code>calculate_ci</code> calculates coverage means across all (regions * design group * bin) combinations. It also estimates a confidence interval for those coverages, using a resampling strategy.</p>
<ul>
<li>
<strong>alpha</strong>: The alpha parameter for the confidence interval calculations.</li>
<li>
<strong>sample_count</strong>: The number of resampling to be performed.</li>
<li>
<strong>resampling_strategy</strong>: The resampling strategy to be used when performing the bootstrap analysis, which can be either ‘profile’ or ‘bin’. In ‘profile’ mode, whole profiles across all bins are resampled. In ‘bin’ mode, each bin is resampled individually and independantly from all others.</li>
</ul>
</div>
<div id="add_metadata" class="section level2">
<h2 class="hasAnchor">
<a href="#add_metadata" class="anchor"></a><code>add_metadata</code>
</h2>
<p><code>add_metadata</code> takes the data-frame produced by <code>calculate_ci</code>and adds region and design metadata to it so it can be more easily plotted.</p>
<ul>
<li>
<strong>design_metadata</strong>: A data-frame providing metadata about the design groups.</li>
</ul>
</div>
<div id="plot" class="section level2">
<h2 class="hasAnchor">
<a href="#plot" class="anchor"></a><code>plot</code>
</h2>
<p>During this step, metagene will use the <code>data.frame</code> provided by àdd_metadata<code>to plot the calculated values using</code>ggplot2`.</p>
<ul>
<li>
<strong>title</strong>: A title for the plot.</li>
<li>
<strong>x_label</strong>: = An x label for the plot.</li>
<li>
<strong>facet_by</strong>:+- A faceting formula.</li>
<li>
<strong>group_by</strong>: Which metadata column should we use for determining the color scale?</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mg$<span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="dt">title =</span> <span class="st">"Demo plot subset"</span>)</code></pre></div>
<p><img src="metagene2_files/figure-html/showPlot-1.png" width="700"></p>
</div>
</div>
<div id="manipulating-the-metagene2-objects" class="section level1">
<h1 class="hasAnchor">
<a href="#manipulating-the-metagene2-objects" class="anchor"></a>Manipulating the <code>metagene2</code> objects</h1>
<div id="getters" class="section level2">
<h2 class="hasAnchor">
<a href="#getters" class="anchor"></a>Getters</h2>
<p>Multiple getters functions are available to access the data that is stored in a <code>metagene2</code> object. Here we present the most relevant ones.</p>
<div id="get_params" class="section level3">
<h3 class="hasAnchor">
<a href="#get_params" class="anchor"></a><code>get_params</code>
</h3>
<p><code>metagene2</code> keeps a list of all analysis parameters used to generate its plots and data structures. This list is initialized with reasonable defaults within the on metagene constructor, and is updated whenever a new parameter is specified in <code>produce_metagene</code>, or any of the sub-processing steps (<code>group_coverages</code>, <code>bin_coverages</code>, etc.). This list can be accessed with the <code>get_params</code> function.</p>
<p>Most sub-processing steps accept NAs for their arguments’ values: in those cases, metagene reuses the last specified value for this parameter, or a default one if the parameter has yet to be defined by the user.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mg &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_demo_metagene.html">get_demo_metagene</a></span>()
<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(mg$<span class="kw">get_params</span>())</code></pre></div>
<pre><code>##  [1] "design"                 "design_metadata"       
##  [3] "bam_files"              "padding_size"          
##  [5] "verbose"                "force_seqlevels"       
##  [7] "paired_end"             "assay"                 
##  [9] "strand_specific"        "paired_end_strand_mode"
## [11] "normalization"          "avoid_gaps"            
## [13] "gaps_threshold"         "bin_count"             
## [15] "alpha"                  "sample_count"          
## [17] "region_mode"            "split_by"              
## [19] "region_filter"          "design_filter"         
## [21] "resampling_strategy"    "facet_by"              
## [23] "group_by"               "title"                 
## [25] "x_label"                "extend_reads"          
## [27] "invert_strand"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mg$<span class="kw">get_params</span>()[<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"bin_count"</span>, <span class="st">"alpha"</span>, <span class="st">"normalization"</span>)]</code></pre></div>
<pre><code>## $bin_count
## [1] 100
## 
## $alpha
## [1] 0.05
## 
## $normalization
## NULL</code></pre>
</div>
<div id="get_bam_count" class="section level3">
<h3 class="hasAnchor">
<a href="#get_bam_count" class="anchor"></a><code>get_bam_count</code>
</h3>
<p>To get the number of aligned reads in a BAM file, you can use the <code>get_bam_count</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mg &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_demo_metagene.html">get_demo_metagene</a></span>()
mg$<span class="kw">get_bam_count</span>(mg$<span class="kw">get_params</span>()[[<span class="st">"bam_files"</span>]][<span class="dv">1</span>])</code></pre></div>
<pre><code>## [1] 1128</code></pre>
</div>
<div id="get_regions" class="section level3">
<h3 class="hasAnchor">
<a href="#get_regions" class="anchor"></a><code>get_regions</code>
</h3>
<p>The <code>get_regions</code> function returns the post-processed regions that <code>metagene2</code> uses internally:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Out demo regions are a GRangesList of two elements containing 50 ranges each.</span>
<span class="kw"><a href="../reference/get_demo_regions.html">get_demo_regions</a></span>()</code></pre></div>
<pre><code>## GRangesList object of length 2:
## $list1
## GRanges object with 50 ranges and 2 metadata columns:
##        seqnames              ranges strand |        name     score
##           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;
##    [1]     chr1   16103663-16105662      + |     list1_1         0
##    [2]     chr1   23921318-23923317      + |     list1_2         0
##    [3]     chr1   34848977-34850976      + |     list1_3         0
##    [4]     chr1   36368182-36370181      + |     list1_4         0
##    [5]     chr1   36690488-36692487      + |     list1_5         0
##    ...      ...                 ...    ... .         ...       ...
##   [46]     chr1 172081530-172083529      + |    list1_46         0
##   [47]     chr1 172081796-172083795      + |    list1_47         0
##   [48]     chr1 172147016-172149015      + |    list1_48         0
##   [49]     chr1 172205805-172207804      + |    list1_49         0
##   [50]     chr1 172260642-172262641      + |    list1_50         0
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths
## 
## $list2
## GRanges object with 50 ranges and 2 metadata columns:
##        seqnames            ranges strand |        name     score
##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;
##    [1]     chr1   3670499-3672498      - |     list2_1         0
##    [2]     chr1   5916399-5918398      - |     list2_2         0
##    [3]     chr1   9699210-9701209      - |     list2_3         0
##    [4]     chr1   9907639-9909638      - |     list2_4         0
##    [5]     chr1 10718946-10720945      - |     list2_5         0
##    ...      ...               ...    ... .         ...       ...
##   [46]     chr1 81075951-81077950      - |    list2_46         0
##   [47]     chr1 85108854-85110853      - |    list2_47         0
##   [48]     chr1 85960056-85962055      - |    list2_48         0
##   [49]     chr1 86110971-86112970      - |    list2_49         0
##   [50]     chr1 87155522-87157521      - |    list2_50         0
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># When we initialize the metagene object, those regions will be pre-processed,</span>
<span class="co"># flattening the list into a single GRanges object and adding a region_name</span>
<span class="co"># column for tracking.</span>
mg &lt;-<span class="st"> </span>metagene2$<span class="kw">new</span>(<span class="dt">regions =</span> <span class="kw"><a href="../reference/get_demo_regions.html">get_demo_regions</a></span>(),
                   <span class="dt">bam_files =</span> <span class="kw"><a href="../reference/get_demo_bam_files.html">get_demo_bam_files</a></span>())

<span class="co"># get_regions allows us to see those post-processed regions.</span>
mg$<span class="kw">get_regions</span>()</code></pre></div>
<pre><code>## GRanges object with 100 ranges and 3 metadata columns:
##         seqnames            ranges strand |        name     score
##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;numeric&gt;
##     [1]     chr1 16103663-16105662      + |     list1_1         0
##     [2]     chr1 23921318-23923317      + |     list1_2         0
##     [3]     chr1 34848977-34850976      + |     list1_3         0
##     [4]     chr1 36368182-36370181      + |     list1_4         0
##     [5]     chr1 36690488-36692487      + |     list1_5         0
##     ...      ...               ...    ... .         ...       ...
##    [96]     chr1 81075951-81077950      - |    list2_46         0
##    [97]     chr1 85108854-85110853      - |    list2_47         0
##    [98]     chr1 85960056-85962055      - |    list2_48         0
##    [99]     chr1 86110971-86112970      - |    list2_49         0
##   [100]     chr1 87155522-87157521      - |    list2_50         0
##         region_name
##         &lt;character&gt;
##     [1]       list1
##     [2]       list1
##     [3]       list1
##     [4]       list1
##     [5]       list1
##     ...         ...
##    [96]       list2
##    [97]       list2
##    [98]       list2
##    [99]       list2
##   [100]       list2
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="get_raw_coverages" class="section level3">
<h3 class="hasAnchor">
<a href="#get_raw_coverages" class="anchor"></a><code>get_raw_coverages</code>
</h3>
<p>To get the coverages produced during the initialization of the <code>metagene2</code> object, you can use the <code>get_raw_coverages</code> function. Please note that to save memory, metagene will only extract the coverages in the selected regions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">coverages &lt;-<span class="st"> </span>mg$<span class="kw">get_raw_coverages</span>()
coverages[[<span class="dv">1</span>]]</code></pre></div>
<pre><code>## RleList of length 22
## $chr10
## numeric-Rle of length 130694993 with 1 run
##   Lengths: 130694993
##   Values :         0
## 
## $chr11
## numeric-Rle of length 122082543 with 1 run
##   Lengths: 122082543
##   Values :         0
## 
## $chr12
## numeric-Rle of length 120129022 with 1 run
##   Lengths: 120129022
##   Values :         0
## 
## $chr13
## numeric-Rle of length 120421639 with 1 run
##   Lengths: 120421639
##   Values :         0
## 
## $chr14
## numeric-Rle of length 124902244 with 1 run
##   Lengths: 124902244
##   Values :         0
## 
## ...
## &lt;17 more elements&gt;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(coverages)</code></pre></div>
<pre><code>## [1] 5</code></pre>
</div>
<div id="get_normalized_coverages" class="section level3">
<h3 class="hasAnchor">
<a href="#get_normalized_coverages" class="anchor"></a><code>get_normalized_coverages</code>
</h3>
<p>The <code>get_normalized_coverages</code> function works exactly like the <code>get_raw_coverages</code> function except that it returns the coverages in read per million aligned (RPM).</p>
</div>
</div>
<div id="clone" class="section level2">
<h2 class="hasAnchor">
<a href="#clone" class="anchor"></a><code>clone</code>
</h2>
<p>To copy a metagene object, you have to use the <code>clone</code> function. Note that certain elements of a metagene object are environments, and will be shared amongst metagene objects unless <code>deep=TRUE</code> is specified.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mg_copy &lt;-<span class="st"> </span>mg$<span class="kw">clone</span>(<span class="dt">deep=</span><span class="ot">TRUE</span>)</code></pre></div>
</div>
</div>
<div id="managing-large-datasets" class="section level1">
<h1 class="hasAnchor">
<a href="#managing-large-datasets" class="anchor"></a>Managing large datasets</h1>
<p>While <code>metagene2</code> tries to reduce its memory usage, it is possible to run into memory limits when working with multiple large datasets (especially when there is a lot of regions with a large width).</p>
<p>One way to avoid this is to analyse each dataset seperately, then merge the resulting data-frames using <code>rbind</code> before passing them to passing them to the <code>plot_metagene</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mg1 &lt;-<span class="st"> </span>metagene2$<span class="kw">new</span>(<span class="dt">bam_files =</span> <span class="kw"><a href="../reference/get_demo_bam_files.html">get_demo_bam_files</a></span>(), 
                     <span class="dt">regions =</span> <span class="kw"><a href="../reference/get_demo_regions.html">get_demo_regions</a></span>()[<span class="dv">1</span>])
mg2 &lt;-<span class="st"> </span>metagene2$<span class="kw">new</span>(<span class="dt">bam_files =</span> <span class="kw"><a href="../reference/get_demo_bam_files.html">get_demo_bam_files</a></span>(),
                     <span class="dt">regions =</span> <span class="kw"><a href="../reference/get_demo_regions.html">get_demo_regions</a></span>()[<span class="dv">2</span>])
<span class="kw"><a href="../reference/plot_metagene.html">plot_metagene</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span>(mg1$<span class="kw">add_metadata</span>(), mg2$<span class="kw">add_metadata</span>()),
              <span class="dt">facet_by=</span>.~region_name)</code></pre></div>
<p><img src="metagene2_files/figure-html/plotMetagene-1.png" width="700"></p>
</div>
<div id="diff_v1" class="section level1">
<h1 class="hasAnchor">
<a href="#diff_v1" class="anchor"></a>Differences with metagene</h1>
<p><code>metagene2</code> is a thorough overhaul of <code>metagene</code>, adding many new features and streamlining others. The most important differences between the two versions are detailed below:</p>
<div id="better-defined-operations" class="section level2">
<h2 class="hasAnchor">
<a href="#better-defined-operations" class="anchor"></a>Better-defined operations</h2>
<p><code>metagene</code> had a 4 step pipeline:</p>
<ol style="list-style-type: decimal">
<li><code>metagene2$new</code></li>
<li><code>produce_table</code></li>
<li><code>produce_data_frame</code></li>
<li><code>plot</code></li>
</ol>
<p>Some of these steps were ambiguous (What is the difference between <code>produce_table</code> and <code>produce_data_frame</code> in this context?) and many useful intermediary structures were hidden from the user. <code>metagene2</code> remedies this by making all intermediary operations explicit:</p>
<ol style="list-style-type: decimal">
<li><code>metagene2$new</code></li>
<li><code>group_coverages</code></li>
<li><code>bin_coverages</code></li>
<li><code>split_coverages_by_regions</code></li>
<li><code>calculate_ci</code></li>
<li><code>add_metadata</code></li>
<li><code>plot</code></li>
</ol>
<p>Each of these intermediary operations now returns the results of their calculations rather than a reference to the <code>metagene2</code> object. As a result, you can no longer chain metagene operations (<code>mg$produce_data_frame()$plot()</code>, for example.) <code>metagene2</code> also adds a new method, <code>produce_metagene</code>, which can be used to go through the whole pipeline all at once.</p>
<p>See the <a href="#intermediary_steps">Intermediary processing steps and further parameters</a> section for more informations on each of the intermediary steps.</p>
</div>
<div id="better-parameter-management" class="section level2">
<h2 class="hasAnchor">
<a href="#better-parameter-management" class="anchor"></a>Better parameter management</h2>
<p>To change certain parameters (<code>alpha</code>, <code>bin_count</code>), <code>metagene</code> required you to call on the intermediary steps directly. Some parameters (like <code>bin_count</code>) had a default value of <code>NA</code>, which allowed you to keep the previous value. Others had a default value (<code>alpha</code>) where <code>NA</code> was an invalid input.</p>
<p>This has been streamlined: now, all intermediary parameters default to <code>NA</code>, and are given reasonable defaults upon object initialization. Also, all parameters can be changed by calling <code>produce_metagene</code>. <code>metagene2</code> manages a smart cache of intermediary results. Upon calling <code>produce_metagene</code> and setting parameters, only the caches of those operations downstream of that parameter are invalidated and recalculated.</p>
</div>
<div id="working-with-metadata" class="section level2">
<h2 class="hasAnchor">
<a href="#working-with-metadata" class="anchor"></a>Working with metadata</h2>
<p>In <code>metagene</code>, the only way to differentiate regions was to split them up in a <code>GRangesList</code> upon object initialization. There was also no way to use information about the various designs/samples when plotting. In our experience, rather than using <code>metagene</code>’s built-in plotting capabilities, most users resorted to getting the <code>data.frame</code> and <code><a href="https://dplyr.tidyverse.org/reference/join.html">dplyr::left_join</a></code>’ing it with their metadata.</p>
<p>In <code>metagene2</code>, metadata management is now part of <code>metagene2</code>object. Regions have explicit metadata (through their <code>mcols</code>, or an explicit <code>region_metadata</code> parameter) which can be used to split them up into metagene units using the <code>split_coverages_by_regions</code> method. Thus, it is no longer necessary to create a new <code>metagene</code> object to change the way regions are grouped together.</p>
<p>Designs/samples can also have metadata, added at the <code>add_metadata</code> step, giving details about antibodies, strains, sample conditions, etc. These information can be used to change the facetting and grouping of the metagenes when plotting, using the <code>facet_by</code>and <code>group_by</code> parameters.</p>
</div>
<div id="true-rna-seq-metagenes" class="section level2">
<h2 class="hasAnchor">
<a href="#true-rna-seq-metagenes" class="anchor"></a>True RNA-seq metagenes</h2>
<p><code>metagene</code> had an experimental “rnaseq” mode. However, this mode mostly generated per-nucleotide coverages over single genes, and did not allow for the aggregations of multiple genes together. Furthermode, <code>metagene</code> did not differentiate between reads that were on different strands, limiting its usefulness for RNA analysis.</p>
<p>In <code>metagene2</code>, rnaseq mode is no longer fundamentally different from chipseq mode. Both allow the binning and aggregation of multiple regions. This behaviour is controlled through the <code>region_mode</code> parameter, which must be set to <code>stitch</code> for the input <code>GRangesList</code> <code>regions</code>object to be treated as a set of exons to be “stitched” together. The <code>strand_specific</code> argument on object initialization also also for reads to be counted only if they lie on the correct strand.</p>
</div>
<div id="miscellaneous-changes-and-improvements" class="section level2">
<h2 class="hasAnchor">
<a href="#miscellaneous-changes-and-improvements" class="anchor"></a>Miscellaneous changes and improvements</h2>
<ul>
<li>
<code>flip_regions</code> and <code>unflip_regions</code> have been removed. The strand is now always taken into account when binning regions.</li>
<li>
<code>add_design</code> has been removed. You can now change the design by setting it directly in <code>groud_coverages</code>or <code>produce_metagene</code>.</li>
<li>
<code>get_matrix</code>, <code>get_table</code> and <code>get_data_frame</code> have been removed. Intermediary results are now accessed by calling the correct step directly (<code>bin_coverages</code> for <code>get_matrix</code>, <code>calculate_ci</code> or <code>add_metadata</code> for <code>get_data_frame</code>)</li>
<li>Parallel processing has been improved! It’s now available on Windows, and it is used in more places where it can provide a performance boost.</li>
<li>
<code>metagene2</code> now supports <code>strand_specific</code>metagenes, with an option to <code>invert_strand</code>s when sequencing cDNA directly.</li>
<li>
<code>metagene2</code> can now <code>extend_reads</code>, a common analysis step when analyzing chip-seq data.</li>
<li>
<code>metagene2</code>now has a <code>plot_single_region</code> method, which can be used to generate a coverage plot for a single region as a diagnostic/QC tool.</li>
<li>The promoters_hg19, promoters_hg18, promoters_mm10 and promoters_mm9 objects have been removed. See the <a href="#common_ranges">Generating common ranges</a> section for details on how to generate equivalent objects yourself.</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li>
<a href="#creating-a-metagene-object">Creating a metagene object</a><ul class="nav nav-pills nav-stacked">
<li><a href="#specifying-alignment-files-bam-files">Specifying alignment files (BAM files)</a></li>
      <li><a href="#specifying-genomic-regions">Specifying genomic regions</a></li>
      </ul>
</li>
      <li>
<a href="#grouping-regions-and-bam-files">Grouping regions and bam files</a><ul class="nav nav-pills nav-stacked">
<li><a href="#grouping-bam-files">Grouping bam files</a></li>
      <li><a href="#grouping-regions">Grouping regions</a></li>
      </ul>
</li>
      <li>
<a href="#intermediary_steps">Intermediary processing steps and further parameters</a><ul class="nav nav-pills nav-stacked">
<li><a href="#arguments-results-caching-and-chaining">Arguments, results caching and chaining</a></li>
      <li><a href="#metagene2new"><code>metagene2$new</code></a></li>
      <li><a href="#group_coverages"><code>group_coverages</code></a></li>
      <li><a href="#bin_coverages"><code>bin_coverages</code></a></li>
      <li><a href="#split_coverages_by_regions"><code>split_coverages_by_regions</code></a></li>
      <li><a href="#calculate_ci"><code>calculate_ci</code></a></li>
      <li><a href="#add_metadata"><code>add_metadata</code></a></li>
      <li><a href="#plot"><code>plot</code></a></li>
      </ul>
</li>
      <li>
<a href="#manipulating-the-metagene2-objects">Manipulating the <code>metagene2</code> objects</a><ul class="nav nav-pills nav-stacked">
<li><a href="#getters">Getters</a></li>
      <li><a href="#clone"><code>clone</code></a></li>
      </ul>
</li>
      <li><a href="#managing-large-datasets">Managing large datasets</a></li>
      <li>
<a href="#diff_v1">Differences with metagene</a><ul class="nav nav-pills nav-stacked">
<li><a href="#better-defined-operations">Better-defined operations</a></li>
      <li><a href="#better-parameter-management">Better parameter management</a></li>
      <li><a href="#working-with-metadata">Working with metadata</a></li>
      <li><a href="#true-rna-seq-metagenes">True RNA-seq metagenes</a></li>
      <li><a href="#miscellaneous-changes-and-improvements">Miscellaneous changes and improvements</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Eric Fournier, Charles Joly Beauparlant, Cedric Lippens, Arnaud Droit.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
